pkg/iter/iterator.go

```diff
@@ -133,38 +135,67 @@ func (i *heapIterator) requeue(ei EntryIterator) {
 }
 
 func (i *heapIterator) Next() bool {
-	if i.curr != nil {
-		i.requeue(i.curr)
-	}
-
 	if i.heap.Len() == 0 {
 		return false
 	}
 
-	i.curr = heap.Pop(i.heap).(EntryIterator)
-	currEntry := i.curr.Entry()
-
-	// keep popping entries off if they match, to dedupe
+	// We support multiple entries with the same timestamp, and we want to
+	// preserve their original order. We look at all the top entries in the
+	// heap with the same timestamp, and pop the ones whose common value
+	// occurs most often.
+	type tuple struct {
+		logproto.Entry
+		EntryIterator
+	}
+	tuples := make([]tuple, 0, i.heap.Len())
 	for i.heap.Len() > 0 {
 		next := i.heap.Peek()
-		nextEntry := next.Entry()
-		if !currEntry.Equal(nextEntry) {
+		entry := next.Entry()
+		if len(tuples) > 0 && !tuples[0].Timestamp.Equal(entry.Timestamp) {
 			break
 		}
 
-		next = heap.Pop(i.heap).(EntryIterator)
-		i.requeue(next)
+		heap.Pop(i.heap)
+		tuples = append(tuples, tuple{
+			Entry:         entry,
+			EntryIterator: next,
+		})
+	}
+
+	// Find in entry which occurs most often which, due to quorum based
```

Comment:
Why do we need to find the entry that occurs most often?   If there are 3 entries with lines "a", "b", "b" then the current intent (thought the current code doesnt do it) is to set currEntry to "b" and requeue "a".   But when Next() is called again "a" will be the only entry for its timestamp and so will be used for currEntry.  

returning "b" first instead of "a" contradicts the earlier comment that when dealing with entries with the same timestamp we preserve the original order.

I dont think that it is possible to preserve the original order when merging chunks, and I also don't feel that it is important.  I think when dealing with entries with the same timestamp we can safely just return them sorted by the content of "Line".   

So, I feel that we can simplify the logic here to just sort the tuple entries by "Line", set currEntry to the first one and requeue all of the others that have a different Line.

eg,
```
	sort.Slice(tuples, func(i, j int) bool {
		return tuples[i].Line < tuples[j].Line
	})
	i.currEntry = tuples[0].Entry

       	// Requeue the iterators, only advancing them if they were not the
	// correct pick.
	for j := range tuples {
		i.requeue(tuples[j].EntryIterator, tuples[j].Line != i.currEntry.Line)
	}
 	return true
}
``` 

Simplifying even more, we could just have `tuples` be a map[string]tuple.  to keep 1 tuple for Entry.Line.  We can then just return the first tuple and requeue the others.  The benefit here is that we are only requeuing 1 Entry per "Line"  which will speed up processing on the following calls to "Next()"

Reply:
Scratch that, i now get how all of this works.

Reply:
:+1: